[21160] Remove rails work_packages views and update specs

oliverguenther opened the pull request:
### TODOS
- [x] Test cukes on their own and remove outdated ones
- [x] Write feature specs for full screen create
### Bugs

(Extend this list with bugs you find)
- [x] Edit work package used to the old view. `/work_packages/:id/edit`  must be defined in angular and should route to the split screen in force edit mode.
- [X] ~~Copy redirects to the `work_package_path` (`/work_packages/:id`) which shows an empty page - the angular router doesn't chime in here.~~ Fixed with https://github.com/oliverguenther/openproject/commit/3ad9ff3a80567d4bb04c3b1220ae3688f22e5f30
- [x] ~~Duplicate work packages used the old view. It 'works' in the sense that no error is thrown but no values are taken over.~~ I have removed the duplicate option for now. The code for it still exists, since we will likely implement it soon enough.

Relevant work package: https://community.openproject.org/work_packages/21160/activity

oliverguenther committed:
Remove work_package route, rename appropriately

oliverguenther committed:
Remove views and actions

oliverguenther committed:
Remove specs

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-features-16364-remove-rails-views.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-features-16364-remove-rails-views.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther committed:
Fix and remove cukes

Removed cukes:
* create.feature: Uses no JS
* copy_with_watchers.feature: Copy feature not implemented
* preview.feature: No preview available
* issue_new.feature: Uses outdated fields no longer shown in create view

oliverguenther commented:
Removed cukes:
- create.feature: Uses no JS
- copy_with_watchers.feature: Copy feature not implemented
- preview.feature: No preview available

oliverguenther committed:
Remove outdated specs

ulferts commented:
The removed actions are still referenced in the permissions definition, e.g.:
- https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L121
- https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L132
- https://github.com/opf/openproject/blob/dev/config/initializers/permissions.rb#L140

(not necessarily complete)

ulferts commented:
There are a whole lot of methods in https://github.com/opf/openproject/blob/dev/app/helpers/work_packages_helper.rb which are no longer necessary I think. Of course, this will not pose immediate problems as this is now code that is no longer referenced, but It will slow down development.

ulferts commented on config/routes.rb:
```diff
@@ -301,9 +301,6 @@
-    # view which is being used for e.g. duplicating work packages.
-    resources :work_packages, only: [:new]
 
     resources :work_packages, only: [:create] do
```
The `:create` action no longer exists, right?

ulferts replied:
There are other references to no longer existing actions in the routes, e.g.:
https://github.com/oliverguenther/openproject/blob/features/16364-remove_rails_views/config/routes.rb#L428

ulferts commented on spec/views/work_package/show.html.erb_spec.rb:
```diff
@@ -37,15 +37,6 @@
-      .to have_selector "img[src='/attachments/#{attachment.id}/download']"
-  end
-
   context 'watchers list is sorted alphabeticaly' do
```
If it is not to much effort, I would move this to somewhere else. It was never a test for work_packages/show to begin with and now this is looking really weird.

oliverguenther replied:
Good idea. Moved to its own `watchers/watchers.html.erb_spec`.

ulferts replied:
Thanks

oliverguenther committed:
Remove other remnants

oliverguenther committed:
Clean up WorkPackagesHelper

oliverguenther committed:
Remove relations

ulferts commented on config/initializers/permissions.rb:
```diff
@@ -127,7 +127,7 @@
+                   { issues: [:edit, :update],
                      :'work_packages/bulk' => [:edit, :update],
                      work_packages: [:edit, :update, :new_type,
                                      :preview, :quoted],
```
What about those here:
`:new_type, :preview, :quoted`
They are not used in the API v2 so they should be removable.

oliverguenther replied:
~~Yup, will do.~~ Done.

ulferts commented on config/routes.rb:
```diff
@@ -427,13 +420,9 @@
+  resources :work_packages, only: [:index] do
     get :column_data, on: :collection # TODO move to API
 
-    resources :relations, controller: 'work_package_relations', only: [:create, :destroy]
```
I think the `WorkPackageRelationsController` can be removed as well.

oliverguenther replied:
Done in updated https://github.com/oliverguenther/openproject/commit/764d54dc698ae0649a5354d8103386dbe6fd67ff

ulferts commented on frontend/app/work_packages/controllers/menus/work-package-context-menu-controller.js:
```diff
@@ -99,6 +99,11 @@ module.exports = function(
+    var selected = getSelectedWorkPackages()[0];
+    return selected && selected.id;
+  }
+
```
You removed this in https://github.com/opf/openproject/pull/3835/files#diff-ab5294bb05227591715f26738a3b98aeL90

ulferts commented on spec/features/work_packages/edit_work_package_spec.rb:
```diff
@@ -0,0 +1,31 @@
+    find('#work-packages--edit-actions-save').click
+
+    subject_field.expect_state_text(subject)
+  end
```
There is a [Page Object](https://github.com/opf/openproject/blob/dev/spec/support/pages/full_work_package.rb) that might help here in addition to the one already used.

oliverguenther replied:
That one only works for the fullscreen, doesn't it? 

The edit route currently leads to the split screen force edit mode. I can extend the test to expand the edit mode to fullscreen, if that's your intention?

ulferts commented:
Nice added vs. removed ratio:

![image](https://cloud.githubusercontent.com/assets/617519/11211666/5dedce54-8d31-11e5-9afd-b04b36bae05c.png)

More of those, please.

oliverguenther commented:
> Nice added vs. removed ratio:

If only the ratio of added vs. removed tests would not be so highly identical to the ratio of removed lines. Oh well, can't have them all. :grimacing:

oliverguenther committed:
Move watchers view spec

oliverguenther committed:
Expect subject after save in edit spec

oliverguenther committed:
Restore report routes

oliverguenther committed:
Restore journals for atom access

oliverguenther committed:
Remove other outdated specs

oliverguenther commented:
The last cuke job cannot be restarted, but this is ready to merge from my side.

ulferts commented:
I'm merging this now and we'll see what our internal CI has to say.
